cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -398,44 +399,44 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
                                         String searchQuery, Set<EntityTypeSimpleName> types,
                                         SortInfo sortInfo, int offset, int limit,
                                         int numCursors, String cursor, boolean showHidden) throws BadRequestException {
-    List<MetadataEntry> results = new ArrayList<>();
-    List<String> cursors = new ArrayList<>();
+    if (offset < 0) {
+      throw new IllegalArgumentException("offset must not be negative");
+    }
+
+    if (limit < 0) {
+      throw new IllegalArgumentException("limit must not be negative");
+    }
+
+    List<MetadataEntry> results = new LinkedList<>();
+    List<String> cursors = new LinkedList<>();
+    List<MetadataEntry> allResults = new LinkedList<>();
     for (MetadataScope scope : scopes) {
       SearchResults searchResults =
         getSearchResults(scope, namespaceId, searchQuery, types, sortInfo, offset, limit, numCursors, cursor,
                          showHidden);
       results.addAll(searchResults.getResults());
       cursors.addAll(searchResults.getCursors());
+      allResults.addAll(searchResults.getAllResults());
```

Comment:
We need to get all of the results so that we can count them in the total, I believe.